plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/ServerPortExtensionTest.java

```diff
@@ -0,0 +1,202 @@
+    shell.setVisible(true);
+    portExtension = new ServerPortExtension();
+    portExtension.createControl(UI_POSITION.BOTTOM, shell);
+    portExtension.setServerWorkingCopy(server);
+  }
+
+  @After
+  public void tearDown() {
+    shell.dispose();
+  }
+
+  @Test
+  public void testCreateControl() {
+    assertNotNull(portExtension.portLabel);
+    assertNotNull(portExtension.portText);
+    assertFalse(portExtension.portLabel.isVisible());
+    assertFalse(portExtension.portText.isVisible());
+  }
+
+  @Test
+  public void testDefaultPort() {
+    assertEquals("8080", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_allowEmpty() {
+    portExtension.portText.setText("");
+    assertTrue(portExtension.portText.getText().isEmpty());
+  }
+
+  @Test
+  public void testPortText_preventNonNumericCharacters() {
+    portExtension.portText.setText("80A80");
+    assertEquals("8080", portExtension.portText.getText());
+
+    portExtension.portText.setText("80 clipboard pasted 80");
+    assertEquals("8080", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_numericCharacters() {
+    portExtension.portText.setText("808");
+    assertEquals("808", portExtension.portText.getText());
+
+    portExtension.portText.setText("8108");
+    assertEquals("8108", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_allowArbitrarilyLargeNumbers() {
+    portExtension.portText.setText("80808");
+    assertEquals("80808", portExtension.portText.getText());
+
+    portExtension.portText.setText("480808");
+    assertEquals("480808", portExtension.portText.getText());
+
+    portExtension.portText.setText("9480808");
+    assertEquals("9480808", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testAttributeSetOnFocusLost() {
+    portExtension.portText.setText("12345");
+    assertEquals("12345", portExtension.portText.getText());
+
+    portExtension.portText.notifyListeners(SWT.FocusOut, null /* event */);
+    verify(server).setAttribute("appEngineDevServerPort", "12345");
+
+    portExtension.portText.setText("54321");
+    assertEquals("54321", portExtension.portText.getText());
+
+    portExtension.portText.notifyListeners(SWT.FocusOut, null /* event */);
+    verify(server).setAttribute("appEngineDevServerPort", "54321");
+  }
+
+  @Test
+  public void testPortText_limitMinimumPortOnFocusLost() {
+    portExtension.portText.setText("0");
+    assertEquals("0", portExtension.portText.getText());
+
+    portExtension.portText.notifyListeners(SWT.FocusOut, null /* event */);
+    assertEquals("1", portExtension.portText.getText());
+  }
+
+  @Test
+  public void testPortText_limitMaximumPortOnFocusLost() {
+    portExtension.portText.setText("123456");
+    assertEquals("123456", portExtension.portText.getText());
+
+    portExtension.portText.notifyListeners(SWT.FocusOut, null /* event */);
+    assertEquals("65535", portExtension.portText.getText());
```

Comment:
I first tried a very similar idea to block users from entering invalid values, such as values exceeding 65535, on the fly. I immediately got a sense that this UX was very wrong. For example, let's assume that a user first entered 10000 and later wanted to change it to 12000. The user could first change it to 120000 by inserting 2, and then delete one '0' to make it 12000. However, this cannot work if we prevent entering 120000.

So, I backed off and consulted the behavior of the `Spinner`. I learned that a user should be given complete freedom on what she can enter as a numeric value — whether she is typing, doing cut-and-paste, or selecting multiple characters. It gives smooth UX; it feels right. If we immediately blocks usual actions such as a key stroke, it doesn't feel right. Lastly, the `Spinner` forces a minimum and a maximum when losing a focus. I followed the same UX.